Temporary page while tljh is building

GeorgianaElena opened the pull request:
This shows a temporary page while tljh is building, that looks like this:
![tljh-is-building-page](https://user-images.githubusercontent.com/7579677/89201201-c9d3fa00-d5b9-11ea-8096-649a7a0d1e7c.gif)

A few things to consider:
* There will be a "down" period between when this page stops showing and JupyterHub starts, because we have to free up the port for traefik.
* The `/logs` page has to be manually refreshed to update the logs
* The main page gets refreshed every 30s so that when JH starts, the login page will be shown instead

TODO
 - [x] Add / update documentation
 - [x] Add / update tests

Closes https://github.com/jupyterhub/the-littlest-jupyterhub/issues/70

yuvipanda commented on bootstrap/bootstrap.py:
```diff
@@ -90,7 +93,38 @@ def validate_host():
+                content = log_file.read()
+                self.wfile.write(content)
+        else:
+            self.path = '/index.html'
```
For routing, I am thinking:

1. if it's '/logs', do what we're doing now
2. if it's for '/', redirect to '/index.html'
3. If it's for '/index.html', serve the file
4. For anything else, return a '403 Forbidden' or '404 Not Found'

We want to only serve the files we want to expose to the public, since otherwise we might accidentally expose secrets to the whole world. A routing approach that explicitly lists routes allowed helps with this.

GeorgianaElena commented:
Thanks for the feedback @yuvipanda :)
I addressed all of the suggestions, except for the password one.

> I also wonder if this should be behind password protection, in case the logs have sensitive information?

There shouldn't be anything sensitive in the logs, sice we're only logging generic messages at each step. I think there might be some in case errors occur, but those go only to `stderr` and are not stored into `installer.log`.

But if we decide to password protect the temporary page, maybe one way would be to use the user and password of the first `tljh` admin user, that are passed to the `bootstrap` script.

What do you think?

yuvipanda commented on bootstrap/index.html:
```diff
@@ -0,0 +1,54 @@
+    <meta name="viewport" content="width=device-width">
+    <img style="width: 150px; height: auto;" src="https://raw.githubusercontent.com/jupyterhub/the-littlest-jupyterhub/master/docs/images/logo/logo.png">
+    <div class="loader center"></div>
+    <div class="center" style="font-size: 30px; font-weight: bold; color: grey; text-align:center;">Please wait while your TLJH is building...</div>
```
Let's put all this under the `<style>` tag and avoid inline styles?

yuvipanda commented on bootstrap/index.html:
```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE html>
+<head>
+  <body>
```
`<body>` should be outside the `<head>`

yuvipanda replied:
We should also add a `<title>` tag

yuvipanda commented on tljh/installer.py:
```diff
@@ -485,6 +486,10 @@ def main():
         help='Plugin pip-specs to install'
     )
+    argparser.add_argument(
+        '--temporary-page',
```
If this is taking the `pid` of the process showing the temporary page, we should rename this to something like `--progress-page-server-pid` or similar? Could also add `type=int` to enforce it's an integer.

yuvipanda commented on tljh/installer.py:
```diff
@@ -499,6 +504,17 @@ def main():
+        try:
+            os.kill(int(args.temporary_page), signal.SIGINT)
+            # Remove the pid file and the temporary html page
+            os.remove('/index.html')
```
We should put everything under `/var/run` or someplace, and not under `/`. This helps keep `/` clean!

yuvipanda commented on bootstrap/bootstrap.py:
```diff
@@ -90,7 +93,47 @@ def validate_host():
+
 def main():
+    temp_page_flag = "--temporary-page"
+    if temp_page_flag in sys.argv:
```
Since `argparse` is in the standard library, what do you think of using that here? Would be cleaner than directly looking in `sys.argv`. However, we do pass all params through to installer, so if it is difficult we should add a comment here on why this isn't using argparse, and let it go :)

GeorgianaElena replied:
I've added a comment about not using `argparse` :sweat_smile:

The reasoning for that was that if using `argparse` here for the temporary page flag, all the other flags should be defined here also (I got complaints about not knowing the other args).  And I think apart from this temp page all of the flags are installer-specific. WDYT?

yuvipanda commented on bootstrap/bootstrap.py:
```diff
@@ -90,7 +93,47 @@ def validate_host():
+    temp_page_flag = "--temporary-page"
+    if temp_page_flag in sys.argv:
+        # Serve the loading page until TLJH builds
+        index_url="https://raw.githubusercontent.com/GeorgianaElena/the-littlest-jupyterhub/in-progress-page/bootstrap/index.html"
```
Hmm, so this means the contents of `index.html` will evolve independently of bootstrap.py. What do you think about embedding the HTML file directly in bootstrap.py?

yuvipanda commented on bootstrap/bootstrap.py:
```diff
@@ -90,7 +93,47 @@ def validate_host():
+            p = multiprocessing.Process(target=serve_forever, args=(loading_page_server,))
+            p.start()
+            # Put the pid after the temp page flag to be passed to the istaller
+            sys.argv.insert(sys.argv.index(temp_page_flag)+1, str(p.pid))
```
We should probably make a copy of sys.argv earlier in `main()` and manipulate that, rather than modify sys.argv directly.

yuvipanda commented:
> But if we decide to password protect the temporary page, maybe one way would be to use the user and password of the first `tljh` admin user, that are passed to the `bootstrap` script.

Yep, that's the way to do it I think! But we can add that as a separate PR, not necessarily part of this one.

yuvipanda commented:
what kinda tests do you think we should add to this?

GeorgianaElena commented:
I added a test in `test_bootrap.py` that checks if there is something at `/index.html` while tljh is building and that the same path returns a 404 once jupyterhub starts.

Do you think this is a good enough test @yuvipanda?

yuvipanda commented on integration-tests/test_bootstrap.py:
```diff
@@ -1,28 +1,36 @@
     ])
+
+    if flags:
+        pkgs = ['python3', 'systemd', 'git']
```
Curious - why do we need git & systemd only if flags is set?

yuvipanda commented on tljh/installer.py:
```diff
@@ -499,6 +505,14 @@ def main():
+        try:
+            os.kill(args.progress_page_server_pid, signal.SIGINT)
+        except Exception as e:
+            pass
```
Let's log an error message when this fails?

yuvipanda commented:
Things look good to me, @GeorgianaElena! I left some minor questions. Once we get the tests to pass this can be merged! \o/

GeorgianaElena commented:
Tests are now passing @yuvipanda, if you want to take another look :tada:

yuvipanda commented on docs/topic/customizing-installer.rst:
```diff
@@ -17,6 +17,35 @@ This page documents the various options you can pass as commandline parameters t
+
+* The page will be accessible at ``http://<tljh-public-ip>/index.html`` in your browser.
+  When TLJH installation is complete, the progress page page will stop and you will be able
+  to access TLJH as usually at ``http://<tljh-public-ip>/index.html``.
```
```suggestion
  to access TLJH as usually at ``http://<tljh-public-ip>``.
```

yuvipanda commented:
The upgrade test is failing thanks to jupyterlab again, so I'm just going to merge this.

AWESOME WORK, @GeorgianaElena! \o/
